File: java/Graph_java/Validate_Binary_Tree_Nodes_1361.java
package java.Graph_java;

import java.util.*;
import java.io.*;

// import static java.lang.Math.max;
// import static java.lang.Math.min;
// import static java.lang.Math.abs;
// import static java.lang.System.out;

public class Validate_Binary_Tree_Nodes_1361 {
    /*
     * Its easy just think of as a directed graph with root node
     * if(thier is cycle || or graph is disconnected (multiple tree with
     * root node))return false
     * else true
     * 
     */
    public boolean dfs(List<List<Integer>> adj, Boolean[] vis, int node) {
        if (node >= adj.size() || node == -1)
            return true;
        if (vis[node] != null)
            return false;
        vis[node] = true;
        for (int i : adj.get(node)) {
            if (dfs(adj, vis, i) == false)
                return false;
        }
        return true;
    }

    public boolean validateBinaryTreeNodes(int n, int[] lc, int[] rc) {
        if (n < 2)
            return true;
        List<List<Integer>> adj = new ArrayList<>();
        Boolean[] vis = new Boolean[n];
        int[] indeg = new int[n];

        for (int i = 0; i < n; ++i) {
            adj.add(new ArrayList<>());
            indeg[i] = 0;
        }
        for (int i = 0; i < n; ++i) {
            if (lc[i] != -1)
                adj.get(i).add(lc[i]);
            if (rc[i] != -1)
                adj.get(i).add(rc[i]);
            if (lc[i] != -1)
                indeg[lc[i]]++;
            if (rc[i] != -1)
                indeg[rc[i]]++;
        }

        for (int i = 0; i < n; ++i) {
            // call dfs where indegree of node is zero and atleast one child
            // why? because 0 indegree can only be root node
            if (indeg[i] == 0 && adj.get(i).size() > 0) {
                if (dfs(adj, vis, i) == false)
                    return false;
                else
                    break;
            }
        }
        for (int i = 0; i < n; ++i) {
            if (vis[i] == null)
                return false;
        }
        return true;
    }

    public static void main(String[] args) throws FileNotFoundException {
        File f = new File("output.txt");

        // Scanner ip = new Scanner(System.in);
        try (Scanner ip = new Scanner(new File("input.txt"));
                PrintWriter op = new PrintWriter(f);) {
            Validate_Binary_Tree_Nodes_1361 obj = new Validate_Binary_Tree_Nodes_1361();
            int nums1[] = { 1, -1, 3, -1 };
            int nums2[] = { 2, -1, -1, -1 };

            op.println(obj.validateBinaryTreeNodes(4, nums1, nums2));

        }
    }

}

File: Code/Graph/Validate_Binary_Tree_Nodes_1361.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
    bool dfs(vector<vector<int>> &adj, vector<bool> &vis, int node)
    {
        if (node >= adj.size() || node == -1)
            return true;
        if (vis[node])
            return false;
        vis[node] = true;
        for (int i : adj[node])
        {
            if (dfs(adj, vis, i) == false)
                return false;
        }
        return true;
    }

    bool validateBinaryTreeNodes(int n, vector<int> lc, vector<int> rc)
    {
        if (n < 2)
            return true;
        vector<vector<int>> adj(n);
        vector<bool> vis(n, false);
        vector<int> indeg(n, 0);

        for (int i = 0; i < n; ++i)
        {
            if (lc[i] != -1)
                adj[i].push_back(lc[i]);
            if (rc[i] != -1)
                adj[i].push_back(rc[i]);
            if (lc[i] != -1)
                indeg[lc[i]]++;
            if (rc[i] != -1)
                indeg[rc[i]]++;
        }

        for (int i = 0; i < n; ++i)
        {
            // call dfs where indegree of node is zero and atleast one child
            // why? because 0 indegree can only be root node
            if (indeg[i] == 0 && adj[i].size() > 0)
            {
                if (dfs(adj, vis, i) == false)
                    return false;
                else
                    break;
            }
        }
        for (int i = 0; i < n; ++i)
        {
            if (vis[i] == false)
                return false;
        }
        return true;
    }
};
int main()
{
#ifndef ONLINE_JUDGE
    freopen("input.txt", "r", stdin);
    freopen("output.txt", "w", stdout);
#endif
    Solution obj;
    auto ans = obj.validateBinaryTreeNodes(4, {1, -1, 3, -1}, {2, -1, -1, -1});
    cout << ans;
    return 0;
}